### ISSUE TITLE
Pass the criteria into beforeUpdate

### ISSUE THREAD

**Issue by cyrilchapon** [balderdashy/waterline#1004]
`beforeUpdate` callback would be much more useful if it carried an `originalObject`, as third argument for example.

Retrieving original object is only achievable by either :

- Providing manually, via [policy](http://stackoverflow.com/a/28747257/4578952) or controller, the id in the request body, which is not acceptable
- [Tweaking](http://stackoverflow.com/a/29644240/4578952) and getting raw JS information on the method `this.update.arguments[0]` to get the PK, then `findOne` manually

**EDIT by @dmarcelino:** Given the interesting discussion generated here, we are re-using this thread for a feature request to add the update `criteria` to the `beforeUpdate` lifecycle in a backwards compatible way:
``` javascript
beforeUpdate: function(valuesToUpdate, cb, criteria){}
```
TL;DR: jump to https://github.com/balderdashy/waterline/issues/1004#issuecomment-101699637

See also #1122

**Comment by dmarcelino** [comment]
Hi @cyrilchapon, this has the performance penalty of requiring another trip to the data store to fetch the original object. Why do you need this and why do you think it's not best to fetch the object in the controller, do what you need, and then call `.update()`?

**Comment by cyrilchapon** [comment]
> this has the performance penalty of requiring another trip to the data store to fetch the original object

Then `originalObjectPK` as a third parameter, leaving the user `findOne`ing the object himself.

> why do you think it's not best to fetch the object in the controller

Because I haven't any controller, I'm using default blueprints actions

> Why do you need this

I'm trying to check if the `myobject.password` variable is a true password update, or just the hash re-sent (According to [this section](http://sailsjs.org/#!/documentation/concepts/ORM/Lifecyclecallbacks.html?q=callbacks-on-update), only "valuesToUpdate" are passed to the `update` method, but that's not always the case, when calling `.save()`. I don't have an example though, I can provide if someone needs it, but I was planning to create another issue for this)

**Comment by dmarcelino** [comment]
I've never used blueprints, is it difficult/complicated to extend/override them?

>  but that's not always the case, when calling `.save()`.

You should be able to do `delete myobject.password;` when you want to avoid `beforeUpdate()` processing the password in a `.save()` call.

Have you considered creating an [instance method](https://github.com/balderdashy/waterline-docs/blob/32e3e8bed3915a043b89c42cc9f97c6a97fdf7d8/models.md#instance-methods) to deal with your case (since you want to avoid creating a controller)?

**Comment by cyrilchapon** [comment]
@dmarcelino ,

> You should be able to do delete myobject.password; when you want to avoid beforeUpdate() processing the password in a .save() call.

I'm not calling `save()` myself, blueprint does. It shouldn't provide a non-updated value but it just does, sometimes. I can't avoid that (and will post an issue about that), and as you said 

> is it difficult/complicated to extend/override them

So I just didn't try this time. Therefore, instead, I'm trying to intercept the action to correct it, and the callback lifecycle seemed a good place for it.

Anyway, about "my case", I just solved it using `this.update.arguments[0]`. My point is not realy to solve my case, since I already solved it. (Thanks for your help though)

My whole point is that wont hurt that a "beforeUpdate" callback provided a reference (and the PK is a good one) to the original object, so we could make some comparison or stuff.

**Comment by dmarcelino** [comment]
> since I already solved it

Good to hear! :grinning: 

> My whole point is that wont hurt that a "beforeUpdate" callback provided a reference (and the PK is a good one) to the original object, so we could make some comparison or stuff.

In my own opinion (probably biased by someone who uses waterline standalone) I'm not entirely convinced as I don't think is worth changing the `.update()` API to pass parameters to `beforeUpdate` nor I think waterline should be doing a `find` before an update. However, I would agree with adding the update `criteria` object to `beforeUpdate`, such as:
``` javascript
beforeUpdate: function(valuesToUpdate, cb, criteria){}
```
This would allow fetching the original values before an update and it wouldn't change `.update()`'s signature. I remember seeing some discussion about this somewhere but I can't find the thread now. @devinivy, do you remember?

**Comment by cyrilchapon** [comment]
@dmarcelino 

> I don't think is worth changing the .update() API

I agree, but I was more thinking of only changing the callback call, to always pass the pk of the object being updated. That would change a waterline API though, and the question "is that worth it" is a good point. I was suggesting a third param for backward compat, but that may be wrong.

> However, I would agree with adding the update criteria object to beforeUpdate

It's interesting but `criteria` can be about **many** objects, whereas the callback concerns only *one instance**. I find there's a point-of-view-gap between you and I on this subject since you seem to see `beforeUpdate` as "before the effective object update called after `update` method", and I more "before any object effective update". Might be difference between our usages, yeah!

**Comment by devinivy** [comment]
I was talking to @particlebanana about this recently.  I think a future version of waterline should pass the criteria as well as the values into beforeUpdate for these sorts of actions.  It would have to be a breaking change, of course!

**Comment by dmarcelino** [comment]
>  I was more thinking of only changing the callback call, to always pass the pk of the object being updated

Sorry, not sure I follow, by "call" you mean the `beforeUpdate` call? For `beforeUpdate` to know the pk of the object being modified it either needs to be passed via param to `.update()` (would require a signature/API change)  or `.update()` would need to do a `.findOne()` to figure out the pk of the object being modified (which would degrade performance).

> It's interesting but criteria can be about many objects

True

> whereas the `beforeUpdate` concerns only one instance.

I don't think so. If you call `.update(criteria, valuesToBeUpdated)` and your query hits several records, `beforeUpdate` is only called once. `beforeUpdate` is called by waterline *a priori* knowing which or how many records will be hit. And this makes sense, otherwise it would be expensive. (someone correct me if I'm wrong)

> I was talking to @particlebanana about this recently.  I think a future version of waterline should pass the criteria as well as the values into beforeUpdate for these sorts of actions.  It would have to be a breaking change, of course!

I thought so, I think that makes sense! :+1:

**Comment by cyrilchapon** [comment]
> I don't think so. If you call `.update(criteria, valuesToBeUpdated)` and your query hits several records, `beforeUpdate` is only called once. beforeUpdate is called by waterline a priori knowing which or how many records will be hit

oO ? Hmmm ok. I was trusting sails [doc](http://sailsjs.org/#!/documentation/concepts/ORM/Lifecyclecallbacks.html) that implies the opposite in [the overview](http://sailsjs.org/#!/documentation/concepts/ORM/Lifecyclecallbacks.html?q=overview) and [the example](http://sailsjs.org/#!/documentation/concepts/ORM/Lifecyclecallbacks.html?q=example)

I'm trying to understand [the process](https://github.com/balderdashy/waterline/search?utf8=%E2%9C%93&q=beforeUpdate) but I'm not able to

**Comment by dmarcelino** [comment]
I've checked your urls and I couldn't find anything that disproved what I mentioned. Let me throw an example to check if we are all in the same page:

Model:
```javascript
module.exports = {
  identity: 'user',
  attributes: {
    user:  'string',
    pass: 'string',
    hashedPass: 'string',
    type: 'string',
  },

  // Lifecycle Callbacks
  beforeUpdate: function (values, cb) {
      values.hashedPass = someHashingFunction(values.pass);
      cb();
    });
  }
};
```

Then we call `.update()`:
```javascript

// DB has the bellow records for User:
// { name: 'joe', pass: '1234', type: 'user' } 
// { name: 'jana', pass: '4321', type: 'user' } 

var criteria = {type: 'user' };
var valuesToUpdate = {pass: 'abcd'};

User.update(criteria, valuesToUpdate, function(){});
```

In this case the criteria hits 2 records and `beforeUpdate` runs only once, resulting in both records sharing the same password and same hash. As far as I know that's how `.beforeUpdate()` works.

**Comment by cyrilchapon** [comment]
Ok, my bad. I got confused by the sails doc and [it's example](http://sailsjs.org/#!/documentation/concepts/ORM/Lifecyclecallbacks.html?q=example). I'm very unfamiliar with raw waterline doc, and I assumed that:

1. If sails.js puts that example, it's because the `beforeCreate` is called once per instance
2. There is no reason that `beforeUpdate`act differently

So the last is just false. Still wondering about the first then.

**Comment by dmarcelino** [comment]
Depends on the inputs, if an input is an array then `beforeCreate can be called more than once:
``` javascript
// beforeCreate called once
User.create({ user: joe }, cb);

// beforeCreate called twice
User.create([{ user: joe }, { user: jane }], cb);
```

**Comment by devinivy** [comment]
There is some confusion there for me– what happens if the adapter has a native `createEach`?

**Comment by dmarcelino** [comment]
I believe (haven't tested) `beforeCreate` is called once before each `createEach` call.

**Comment by cyrilchapon** [comment]
> I believe (haven't tested) beforeCreate is called once before each createEach() call.

The exact behavior I was assuming and expecting from beforeUpdate

**Comment by devinivy** [comment]
That's not really fair, though– in the case of `createEach`, records are provided by the query itself.  To know what records an update is going to affect, an extra query would need to be made.  And it may be out-of-date by the time the update is performed anyway!  It seems pretty clear to me that the solution is to pass both criteria and values to `beforeUpdate` so that a user can optionally perform the `find` you'd like to do.  I think what you're asking to able to achieve is very reasonable, but we can't bake the overhead of that `find` into waterline core.

**Comment by dmarcelino** [comment]
Waterline is a *lightish* interface between your code and DB calls. So, when you do:
```js
User.update( {type: 'user' }, {pass: 'abcd'}, function(){});
```

waterline instructs the DB, SQL for example:
```
UPDATE table_user
SET pass = 'abcd'
WHERE type = 'user' ;
```

Before the `UPDATE` query comes through waterline (and `beforeUpdate`) has no idea which records will be modified.

**Comment by cyrilchapon** [comment]
> in the case of createEach, records are provided by the query itself  
> To know what records an update is going to affect, an extra query would need to be made

@devinivy Yep I got it know. That's something transparent when you're using sails.js' blueprint APIs.

> Before the UPDATE query comes through waterline (and beforeUpdate) has no idea which records will be modified.

> I think what you're asking to able to achieve is very reasonable, but we can't bake the overhead of that find into waterline core.

@devinivy @dmarcelino Yes ok. That's why `criteria` would be achievable and `originalObject` wouldn't.

**Comment by dmarcelino** [comment]
BTW, if you did (I don't think it's currently working - balderdashy/waterline-docs#67) something like:

```javascript
var criteria1 = {name: 'joe' };
var criteria2 = {name: 'jane' };
var valuesToUpdate1 = {pass: 'abcd'};
var valuesToUpdate2 = {pass: 'dcba'};

User.update([criteria1, criteria2], [valuesToUpdate1, valuesToUpdate2], function(){});
```

Then `beforeUpdate` would ran twice. Same as `beforeCreate` when the param is an array. In other words the API does try to be consistent. Unfortunately I think there is a bug around this or it was never implemented.

> That's why `criteria` would be achievable and `originalObject` wouldn't.

Yeap.

**Comment by dmarcelino** [comment]
> I was talking to @particlebanana about this recently. I think a future version of waterline should pass the criteria as well as the values into beforeUpdate for these sorts of actions. It would have to be a breaking change, of course!

@devinivy, is there any issue opened for that? If not I think we can re-use this one with a title change.

**Comment by devinivy** [comment]
No issue that I know of!

**Comment by dmarcelino** [comment]
If someone volunteers I think this could be a good fit for `0.11`.

**Comment by devinivy** [comment]
Oops!  Slip of the hand.  Meant to say that I'd be happy to do this, or to facilitate.  I'll start by pointing-out some lines of code:
https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L50
https://github.com/balderdashy/waterline/blob/master/lib/waterline/query/dql/update.js#L156-L172

**Comment by cyrilchapon** [comment]
Also note, for community maybe looking for a temporary solution, that `this.update.arguments[0]` inside `beforeUpdate` update returns either

- the id (as a string) (if one calls `PUT /resource/1` with data, for example)
- or sometimes the criteria (as an object) (if one calls waterline raw `update()` with criteria)

**Comment by leobudima** [comment]
Does anyone know why suddenly `this.update.arguments` started returning `null`? Broke my app after running `npm update`...

**Comment by devinivy** [comment]
Which version were you on, and when did you npm update?  We did push a new version about two hours ago.  That said, I don't see any changes that would change what you're talking about.  Also, I don't think there's any official support / tests for using `this.update.arguments`.  I've never seen it used!  In this case, does `this` refer to the collection?

**Comment by devinivy** [comment]
Could this be an update to your version of node / io tied to V8?  I see this note on MDN, at least: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/arguments

**Comment by leobudima** [comment]
Hi @devinivy - thanks a lot for replying! I've updated to 0.10.24, but I really can't tell (stupidly) what was the previous version...

`this` still points to the collection. 

I don't think it's related to node/io update since it got broken first on Heroku, and then, when I ran `npm update`, on my development machine.

Investigating further and thanks a lot for the assistance!

**Comment by leobudima** [comment]
As I pushed an update to Heroku yesterday (when it runs `npm install`), and it went without problems, it must be a recent change to one of the packages.

**Comment by devinivy** [comment]
Sure thing!  The only other idea I have is some change made by async in our upgrade from 1.0.0 to 1.2.0.  I can't find anything in waterline that would set that property manually.

**Comment by leobudima** [comment]
That was a good idea - and there's the reason. When I revert back to 1.0.0, it works again :)

**Comment by leobudima** [comment]
I can confirm that, with async 1.1.0+,  `this.update.arguments` approach no longer works. Something in their code has changed arguments that are passed around in `series` and underlying functions.

Thanks again for pointing me into right direction, @devinivy.

**Comment by zinas** [comment]
Any plans on fixing this? (I mean the async dep making this.update.arguments not work)

**Comment by mmiller42** [comment]
Just FYI to those using Sails and trying to find a good way to downgrade Async to use this `update.arguments` hack but without modifying your dependencies *too* much:

1. `cd /path/to/your/project/root`
1. `npm install sails@0.11.0` (the last version of Sails with minimum Waterline dependency starting before 0.10.24)
2. `cd node_modules/sails`
3. `npm install waterline@0.10.23` (the last version of Waterline with `"async": ~1.0.0`)
4. `cd node_modules/waterline`
5. `npm install async@1.0.0`
6. `cd ../../../..` (up to project root)
7. `npm shrinkwrap` (freezes your dependency versions and their sub-dependencies)
8. Commit `npm-shrinkwrap.json` so other team members are sure to install Async at version 1.0.0

Sort of crappy since it freezes *all* dependencies, but hopefully this will be a temporary solution and we'll be able to access criteria another way in a future version of Waterline.

**Comment by mmiller42** [comment]
Additionally, I've created a pull request #1122 if anyone in this thread is interested in reviewing, for the 0.11 release. It is a breaking change that would modify the lifecycle callbacks to add the criteria as one of the arguments, i.e. `beforeUpdate: function (values, criteria, next)` and `beforeValidate: function (values, criteria, next)` (in the latter, `criteria` is `null` when `beforeValidate` is called during the create operation).

**Comment by pfzero** [comment]
Link to this [SO question](http://stackoverflow.com/questions/32205654/pass-criteria-to-validation-and-lifecycle-callbacks-in-sailsjs) which will basically will have an answer when this issue is solved.

**Comment by sailsbot** [comment]
Thanks for posting, @cyrilchapon.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by mmiller42** [comment]
@cyrilchapon, @devinivy, @dmarcelino 
If any of you would like to review PR #1122, I'd really appreciate it. Right now my own blog app is relying on this fork.

In summary, it adds a criteria parameter to the invocations of the beforeUpdate, beforeDestroy, and beforeValidate callbacks. For backwards compatibility, it detects if the callbacks's `length` is 3 (or 2 for beforeDestroy) and, if so, it does not pass criteria, so this should be a non-breaking change.

It'd be awesome if this could make it into Waterline 0.11.

**Comment by cyrilchapon** [comment]
@mmiller42, thanks about this.
I'm pretty far away from sails.js these days and wont have this opportunity soon, but thanks for this man. I've been struggling so much with that, this is great news.

**Comment by sailsbot** [comment]
Thanks for posting, @cyrilchapon.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by LeonardoGentile** [comment]
+1

**Comment by sailsbot** [comment]
Thanks for posting, @cyrilchapon.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by devinivy** [comment]
Open as part of an active PR #1122.

**Comment by particlebanana** [comment]
This is on the roadmap and accepted into the `0.11` release. The PR is open and contains a valid solution. Closing this and moving the conversation to the PR.
